From 495d619b64600a1f76ea0e73c333287c105a3d30 Mon Sep 17 00:00:00 2001
From: Chih-Wei Huang <cwhuang@linux.org.tw>
Date: Thu, 1 Jun 2017 12:00:10 +0800
Subject: [PATCH 13/27] media: enable dithering for RGB565 conversion

This seems to improve video playback quality.

Inspired by a patch from WuZhen.
---
 .../colorconversion/ColorConverter.cpp          | 17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/media/libstagefright/colorconversion/ColorConverter.cpp b/media/libstagefright/colorconversion/ColorConverter.cpp
index 4b4f65f13c..bd786a7df5 100644
--- a/media/libstagefright/colorconversion/ColorConverter.cpp
+++ b/media/libstagefright/colorconversion/ColorConverter.cpp
@@ -17,6 +17,7 @@
 //#define LOG_NDEBUG 0
 #define LOG_TAG "ColorConverter"
 #include <android-base/macros.h>
+#include <cutils/properties.h>
 #include <utils/Log.h>
 
 #include <media/stagefright/foundation/ADebug.h>
@@ -575,9 +576,19 @@ status_t ColorConverter::convertYUV420PlanarUseLibYUV(
     switch (mDstFormat) {
     case OMX_COLOR_Format16bitRGB565:
     {
-        DECLARE_YUV2RGBFUNC(func, RGB565);
-        (*func)(src_y, src.mStride, src_u, src.mStride / 2, src_v, src.mStride / 2,
-                (uint8_t *)dst_ptr, dst.mStride, src.cropWidth(), src.cropHeight());
+        char property[PROPERTY_VALUE_MAX];
+        bool isMesa = false;
+        if (property_get("ro.hardware.egl", property, "default") > 0)
+            if (strcmp(property, "mesa") == 0)
+                isMesa = true;
+        if (isMesa) {
+            libyuv::I420ToRGB565Dither(src_y, src.mStride, src_u, src.mStride / 2, src_v, src.mStride / 2,
+                    (uint8_t *)dst_ptr, dst.mStride, NULL, src.cropWidth(), src.cropHeight());
+        } else {
+            DECLARE_YUV2RGBFUNC(func, RGB565);
+            (*func)(src_y, src.mStride, src_u, src.mStride / 2, src_v, src.mStride / 2,
+                    (uint8_t *)dst_ptr, dst.mStride, src.cropWidth(), src.cropHeight());
+        }
         break;
     }
 
-- 
2.40.0

